From 8335dbaacb1fc136747a736e1b476f9aaf77cd56 Mon Sep 17 00:00:00 2001
From: Sparks Song <shijiao@amazon.com>
Date: Wed, 26 Aug 2026 23:16:35 +0000
Subject: [PATCH] Revert "selinux: fix overlayfs mmap() and mprotect() access
 checks"

This reverts commit d844702198395d3f80222777030f69db6be6b709.

[vyaghras:
Ported to work with the changes introduced in
d127c7e1f3e1231d8ac8b2738e3f00bea12afe74 and 
9d99b770e7b67b00bdef9005b928aad13e1d679b.
]
Signed-off-by: Shikha vyaghra <vyaghras@amazon.com>
---
 security/selinux/hooks.c          | 388 ++++++------------------------
 security/selinux/include/objsec.h |  19 --
 2 files changed, 76 insertions(+), 331 deletions(-)

diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index eef75bde6..5efd44467 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -1665,32 +1665,26 @@ static int cred_has_capability(const struct cred *cred,
 	return rc;
 }
 
-/*
- * Check whether a SID has a particular permission to an inode.  The 'adp'
- * parameter is optional and allows other audit data to be passed (e.g. the
- * dentry).
- */
-static int inode_sid_has_perm(u32 sid, struct inode *inode, u32 perms,
-			      struct common_audit_data *adp)
+/* Check whether a task has a particular permission to an inode.
+   The 'adp' parameter is optional and allows other audit
+   data to be passed (e.g. the dentry). */
+static int inode_has_perm(const struct cred *cred,
+			  struct inode *inode,
+			  u32 perms,
+			  struct common_audit_data *adp)
 {
 	struct inode_security_struct *isec;
+	u32 sid;
 
 	if (unlikely(IS_PRIVATE(inode)))
 		return 0;
 
+	sid = cred_sid(cred);
 	isec = selinux_inode(inode);
 
 	return avc_has_perm(sid, isec->sid, isec->sclass, perms, adp);
 }
 
-static int inode_has_perm(const struct cred *cred,
-			  struct inode *inode,
-			  u32 perms,
-			  struct common_audit_data *adp)
-{
-	return inode_sid_has_perm(cred_sid(cred), inode, perms, adp);
-}
-
 /* Same as inode_has_perm, but pass explicit audit data containing
    the dentry to help the auditing code to more easily generate the
    pathname if needed. */
@@ -1745,72 +1739,49 @@ static inline int file_path_has_perm(const struct cred *cred,
 static int bpf_fd_pass(const struct file *file, u32 sid);
 #endif
 
-static int __file_has_perm(const struct cred *cred, const struct file *file,
-			   u32 av, bool bf_user_file)
-
+/* Check whether a task can use an open file descriptor to
+   access an inode in a given way.  Check access to the
+   descriptor itself, and then use dentry_has_perm to
+   check a particular permission to the file.
+   Access to the descriptor is implicitly granted if it
+   has the same SID as the process.  If av is zero, then
+   access to the file is not checked, e.g. for cases
+   where only the descriptor is affected like seek. */
+static int file_has_perm(const struct cred *cred,
+			 struct file *file,
+			 u32 av)
 {
+	struct file_security_struct *fsec = selinux_file(file);
+	struct inode *inode = file_inode(file);
 	struct common_audit_data ad;
-	struct inode *inode;
-	u32 ssid = cred_sid(cred);
-	u32 tsid_fd;
+	u32 sid = cred_sid(cred);
 	int rc;
 
-	if (bf_user_file) {
-		struct backing_file_security_struct *bfsec;
-		const struct path *path;
-
-		if (WARN_ON(!(file->f_mode & FMODE_BACKING)))
-			return -EIO;
-
-		bfsec = selinux_backing_file(file);
-		path = backing_file_user_path(file);
-		tsid_fd = bfsec->uf_sid;
-		inode = d_inode(path->dentry);
-
-		ad.type = LSM_AUDIT_DATA_PATH;
-		ad.u.path = *path;
-	} else {
-		struct file_security_struct *fsec = selinux_file(file);
-
-		tsid_fd = fsec->sid;
-		inode = file_inode(file);
-
-		ad.type = LSM_AUDIT_DATA_FILE;
-		ad.u.file = file;
-	}
+	ad.type = LSM_AUDIT_DATA_FILE;
+	ad.u.file = file;
 
-	if (ssid != tsid_fd) {
-		rc = avc_has_perm(ssid, tsid_fd, SECCLASS_FD, FD__USE, &ad);
+	if (sid != fsec->sid) {
+		rc = avc_has_perm(sid, fsec->sid,
+				  SECCLASS_FD,
+				  FD__USE,
+				  &ad);
 		if (rc)
-			return rc;
+			goto out;
 	}
 
 #ifdef CONFIG_BPF_SYSCALL
-	/* regardless of backing vs user file, use the underlying file here */
-	rc = bpf_fd_pass(file, ssid);
+	rc = bpf_fd_pass(file, cred_sid(cred));
 	if (rc)
 		return rc;
 #endif
 
 	/* av is zero if only checking access to the descriptor. */
+	rc = 0;
 	if (av)
-		return inode_has_perm(cred, inode, av, &ad);
-
-	return 0;
-}
+		rc = inode_has_perm(cred, inode, av, &ad);
 
-/* Check whether a task can use an open file descriptor to
-   access an inode in a given way.  Check access to the
-   descriptor itself, and then use dentry_has_perm to
-   check a particular permission to the file.
-   Access to the descriptor is implicitly granted if it
-   has the same SID as the process.  If av is zero, then
-   access to the file is not checked, e.g. for cases
-   where only the descriptor is affected like seek. */
-static inline int file_has_perm(const struct cred *cred,
-				const struct file *file, u32 av)
-{
-	return __file_has_perm(cred, file, av, false);
+out:
+	return rc;
 }
 
 /*
@@ -3828,74 +3799,6 @@ static int selinux_file_alloc_security(struct file *file)
 	return 0;
 }
 
-static inline u32 selinux_file_user_sid(const struct file *file)
-{
-	if (unlikely(file->f_mode & FMODE_BACKING))
-		return selinux_backing_file(file)->uf_sid;
-	return selinux_file(file)->sid;
-}
-
-static int selinux_backing_file_alloc(struct file *backing_file,
-				      const struct file *user_file)
-{
-	struct backing_file_security_struct *bfsec;
-	const struct backing_file_security_struct *ubfsec;
-	struct backing_file_security_layer *layer;
-	u32 i;
-
-	bfsec = selinux_backing_file(backing_file);
-	bfsec->uf_sid = selinux_file_user_sid(user_file);
-	if (!(user_file->f_mode & FMODE_BACKING))
-		return 0;
-
-	ubfsec = selinux_backing_file(user_file);
-	/* a wrapped count would make kmalloc_array() return ZERO_SIZE_PTR */
-	if (unlikely(ubfsec->layer_count == U32_MAX))
-		return -EOVERFLOW;
-
-	/*
-	 * The final VMA only retains the lowest backing file, so record the
-	 * whole chain here rather than in the mmap hook, where concurrent
-	 * mappings would have to be serialized.  Size it dynamically: erofs
-	 * inode sharing adds a backing file without bumping s_stack_depth.
-	 */
-	bfsec->layers = kmalloc_array(ubfsec->layer_count + 1,
-				      sizeof(*bfsec->layers), GFP_KERNEL);
-	if (!bfsec->layers)
-		return -ENOMEM;
-
-	for (i = 0; i < ubfsec->layer_count; i++) {
-		layer = &bfsec->layers[i];
-		*layer = ubfsec->layers[i];
-		path_get(&layer->path);
-	}
-
-	/* f_path, not file_user_path(): this layer, not the top-level file */
-	layer = &bfsec->layers[i];
-	layer->path = user_file->f_path;
-	layer->mounter_sid = cred_sid(user_file->f_cred);
-	layer->fd_sid = selinux_file(user_file)->sid;
-	path_get(&layer->path);
-	bfsec->layer_count = ubfsec->layer_count + 1;
-
-	return 0;
-}
-
-static void selinux_backing_file_free(struct file *backing_file)
-{
-	struct backing_file_security_struct *bfsec;
-
-	/* security_backing_file_free() may be called twice after an error */
-	if (!backing_file_security(backing_file))
-		return;
-
-	bfsec = selinux_backing_file(backing_file);
-	while (bfsec->layer_count)
-		path_put(&bfsec->layers[--bfsec->layer_count].path);
-	kfree(bfsec->layers);
-	bfsec->layers = NULL;
-}
-
 /*
  * Check whether a task has the ioctl permission and cmd
  * operation to an inode.
@@ -4013,100 +3916,42 @@ static int selinux_file_ioctl_compat(struct file *file, unsigned int cmd,
 
 static int default_noexec __ro_after_init;
 
-static u32 file_map_prot_to_av(unsigned long prot, bool shared)
-{
-	u32 av = FILE__READ;
-
-	if (shared && (prot & PROT_WRITE))
-		av |= FILE__WRITE;
-	if (prot & PROT_EXEC)
-		av |= FILE__EXECUTE;
-
-	return av;
-}
-
-static int backing_mounters_has_perm(const struct file *file, u32 av)
-{
-	const struct backing_file_security_struct *bfsec;
-	const struct backing_file_security_layer *layer;
-	struct common_audit_data ad;
-	struct inode *inode;
-	u32 i;
-	int rc;
-
-	if (WARN_ON_ONCE(!(file->f_mode & FMODE_BACKING)))
-		return -EIO;
-
-	bfsec = selinux_backing_file(file);
-	for (i = 0; i < bfsec->layer_count; i++) {
-		layer = &bfsec->layers[i];
-		inode = d_inode(layer->path.dentry);
-
-		ad.type = LSM_AUDIT_DATA_PATH;
-		ad.u.path = layer->path;
-
-		if (layer->mounter_sid != layer->fd_sid) {
-			rc = avc_has_perm(layer->mounter_sid, layer->fd_sid,
-					  SECCLASS_FD, FD__USE, &ad);
-			if (rc)
-				return rc;
-		}
-
-		rc = inode_sid_has_perm(layer->mounter_sid, inode, av, &ad);
-		if (rc)
-			return rc;
-	}
-
-	return 0;
-}
-
-static int __file_map_prot_check(const struct file *file, unsigned long prot,
-				 bool shared, bool mounter_check,
-				 bool bf_user_file)
+static int file_map_prot_check(struct file *file, unsigned long prot, int shared)
 {
-	struct inode *inode = NULL;
-	bool prot_exec = prot & PROT_EXEC;
-	bool prot_write = prot & PROT_WRITE;
-
-	if (file) {
-		if (bf_user_file)
-			inode = d_inode(backing_file_user_path(file)->dentry);
-		else
-			inode = file_inode(file);
-	}
-
-	if (!mounter_check && default_noexec && prot_exec &&
-	    (!file || IS_PRIVATE(inode) || (!shared && prot_write))) {
-		int rc;
-		u32 sid = current_sid();
+	const struct cred *cred = current_cred();
+	u32 sid = cred_sid(cred);
+	int rc = 0;
 
+	if (default_noexec &&
+	    (prot & PROT_EXEC) && (!file || IS_PRIVATE(file_inode(file)) ||
+				   (!shared && (prot & PROT_WRITE)))) {
 		/*
-		 * We are making executable an anonymous mapping or a private
-		 * file mapping that will also be writable.
+		 * We are making executable an anonymous mapping or a
+		 * private file mapping that will also be writable.
+		 * This has an additional check.
 		 */
-		rc = avc_has_perm(sid, sid, SECCLASS_PROCESS, PROCESS__EXECMEM,
-				  NULL);
+		rc = avc_has_perm(sid, sid, SECCLASS_PROCESS,
+				  PROCESS__EXECMEM, NULL);
 		if (rc)
-			return rc;
+			goto error;
 	}
 
 	if (file) {
-		const struct cred *cred = mounter_check ?
-				file->f_cred : current_cred();
+		/* read access is always possible with a mapping */
+		u32 av = FILE__READ;
 
-		return __file_has_perm(cred, file,
-				       file_map_prot_to_av(prot, shared),
-				       bf_user_file);
-	}
+		/* write access only matters if the mapping is shared */
+		if (shared && (prot & PROT_WRITE))
+			av |= FILE__WRITE;
 
-	return 0;
-}
+		if (prot & PROT_EXEC)
+			av |= FILE__EXECUTE;
 
-static inline int file_map_prot_check(const struct file *file,
-				      unsigned long prot, bool shared,
-				      bool mounter_check)
-{
-	return __file_map_prot_check(file, prot, shared, mounter_check, false);
+		return file_has_perm(cred, file, av);
+	}
+
+error:
+	return rc;
 }
 
 static int selinux_mmap_addr(unsigned long addr)
@@ -4122,85 +3967,36 @@ static int selinux_mmap_addr(unsigned long addr)
 	return rc;
 }
 
-static int selinux_mmap_file_common(struct file *file, unsigned long prot,
-				    bool shared, bool mounter_check)
+static int selinux_mmap_file(struct file *file,
+			     unsigned long reqprot __always_unused,
+			     unsigned long prot, unsigned long flags)
 {
-	if (file) {
-		int rc;
-		struct common_audit_data ad;
-		const struct cred *cred = mounter_check ?
-				file->f_cred : current_cred();
+	struct common_audit_data ad;
+	int rc;
 
+	if (file) {
 		ad.type = LSM_AUDIT_DATA_FILE;
 		ad.u.file = file;
-		rc = inode_has_perm(cred, file_inode(file), FILE__MAP, &ad);
+		rc = inode_has_perm(current_cred(), file_inode(file),
+				    FILE__MAP, &ad);
 		if (rc)
 			return rc;
 	}
 
-	return file_map_prot_check(file, prot, shared, mounter_check);
-}
-
-static int selinux_mmap_file(struct file *file,
-			     unsigned long reqprot __always_unused,
-			     unsigned long prot, unsigned long flags)
-{
-	return selinux_mmap_file_common(file, prot,
-					(flags & MAP_TYPE) == MAP_SHARED,
-					false);
-}
-
-/**
- * selinux_mmap_backing_file - Check mmap permissions on a backing file
- * @vma: memory region
- * @backing_file: stacked filesystem backing file
- * @user_file: user visible file
- *
- * This is called after selinux_mmap_file() on stacked filesystems, and it
- * is this function's responsibility to verify access to @backing_file and
- * setup the SELinux state for possible later use in the mprotect() code path.
- *
- * By the time this function is called, mmap() access to @user_file has already
- * been authorized and @vma->vm_file has been set to point to @backing_file.
- *
- * Return zero on success, negative values otherwise.
- */
-static int selinux_mmap_backing_file(struct vm_area_struct *vma,
-				     struct file *backing_file,
-				     struct file *user_file __always_unused)
-{
-	unsigned long prot = 0;
-
-	/* translate vma->vm_flags perms into PROT perms */
-	if (vma->vm_flags & VM_READ)
-		prot |= PROT_READ;
-	if (vma->vm_flags & VM_WRITE)
-		prot |= PROT_WRITE;
-	if (vma->vm_flags & VM_EXEC)
-		prot |= PROT_EXEC;
-
-	return selinux_mmap_file_common(backing_file, prot,
-					vma->vm_flags & VM_SHARED,
-					true);
+	return file_map_prot_check(file, prot,
+				   (flags & MAP_TYPE) == MAP_SHARED);
 }
 
 static int selinux_file_mprotect(struct vm_area_struct *vma,
 				 unsigned long reqprot __always_unused,
 				 unsigned long prot)
 {
-	int rc;
 	const struct cred *cred = current_cred();
 	u32 sid = cred_sid(cred);
-	u32 av;
-	const struct file *file = vma->vm_file;
-	bool backing_file;
-	bool shared = vma->vm_flags & VM_SHARED;
-
-	/* check if we need to trigger the "backing files are awful" mode */
-	backing_file = file && (file->f_mode & FMODE_BACKING);
 
 	if (default_noexec &&
 	    (prot & PROT_EXEC) && !(vma->vm_flags & VM_EXEC)) {
+		int rc = 0;
 		/*
 		 * We don't use the vma_is_initial_heap() helper as it has
 		 * a history of problems and is currently broken on systems
@@ -4214,15 +4010,11 @@ static int selinux_file_mprotect(struct vm_area_struct *vma,
 		    vma->vm_end <= vma->vm_mm->brk) {
 			rc = avc_has_perm(sid, sid, SECCLASS_PROCESS,
 					  PROCESS__EXECHEAP, NULL);
-			if (rc)
-				return rc;
-		} else if (!file && (vma_is_initial_stack(vma) ||
+		} else if (!vma->vm_file && (vma_is_initial_stack(vma) ||
 			    vma_is_stack_for_current(vma))) {
 			rc = avc_has_perm(sid, sid, SECCLASS_PROCESS,
 					  PROCESS__EXECSTACK, NULL);
-			if (rc)
-				return rc;
-		} else if (file && vma->anon_vma) {
+		} else if (vma->vm_file && vma->anon_vma) {
 			/*
 			 * We are making executable a file mapping that has
 			 * had some COW done. Since pages might have been
@@ -4230,37 +4022,13 @@ static int selinux_file_mprotect(struct vm_area_struct *vma,
 			 * modified content.  This typically should only
 			 * occur for text relocations.
 			 */
-			rc = __file_has_perm(cred, file, FILE__EXECMOD,
-					     backing_file);
-			if (rc)
-				return rc;
-			if (backing_file) {
-				rc = backing_mounters_has_perm(file,
-							       FILE__EXECMOD);
-				if (rc)
-					return rc;
-				rc = file_has_perm(file->f_cred, file,
-						   FILE__EXECMOD);
-				if (rc)
-					return rc;
-			}
+			rc = file_has_perm(cred, vma->vm_file, FILE__EXECMOD);
 		}
-	}
-
-	rc = __file_map_prot_check(file, prot, shared, false, backing_file);
-	if (rc)
-		return rc;
-	if (backing_file) {
-		av = file_map_prot_to_av(prot, shared);
-		rc = backing_mounters_has_perm(file, av);
-		if (rc)
-			return rc;
-		rc = file_map_prot_check(file, prot, shared, true);
 		if (rc)
 			return rc;
 	}
 
-	return 0;
+	return file_map_prot_check(vma->vm_file, prot, vma->vm_flags&VM_SHARED);
 }
 
 static int selinux_file_lock(struct file *file, unsigned int cmd)
@@ -7390,7 +7158,6 @@ struct lsm_blob_sizes selinux_blob_sizes __ro_after_init = {
 	.lbs_cred = sizeof(struct cred_security_struct),
 	.lbs_task = sizeof(struct task_security_struct),
 	.lbs_file = sizeof(struct file_security_struct),
-	.lbs_backing_file = sizeof(struct backing_file_security_struct),
 	.lbs_inode = sizeof(struct inode_security_struct),
 	.lbs_ipc = sizeof(struct ipc_security_struct),
 	.lbs_key = sizeof(struct key_security_struct),
@@ -7614,12 +7381,9 @@ static struct security_hook_list selinux_hooks[] __ro_after_init = {
 
 	LSM_HOOK_INIT(file_permission, selinux_file_permission),
 	LSM_HOOK_INIT(file_alloc_security, selinux_file_alloc_security),
-	LSM_HOOK_INIT(backing_file_alloc, selinux_backing_file_alloc),
-	LSM_HOOK_INIT(backing_file_free, selinux_backing_file_free),
 	LSM_HOOK_INIT(file_ioctl, selinux_file_ioctl),
 	LSM_HOOK_INIT(file_ioctl_compat, selinux_file_ioctl_compat),
 	LSM_HOOK_INIT(mmap_file, selinux_mmap_file),
-	LSM_HOOK_INIT(mmap_backing_file, selinux_mmap_backing_file),
 	LSM_HOOK_INIT(mmap_addr, selinux_mmap_addr),
 	LSM_HOOK_INIT(file_mprotect, selinux_file_mprotect),
 	LSM_HOOK_INIT(file_lock, selinux_file_lock),
diff --git a/security/selinux/include/objsec.h b/security/selinux/include/objsec.h
index bbb04f26b..816fde5a5 100644
--- a/security/selinux/include/objsec.h
+++ b/security/selinux/include/objsec.h
@@ -86,18 +86,6 @@ struct file_security_struct {
 	u32 pseqno; /* Policy seqno at the time of file open */
 };
 
-struct backing_file_security_layer {
-	struct path path; /* this layer's real path */
-	u32 mounter_sid; /* SID of the mounter that opened it */
-	u32 fd_sid; /* SID of its open file description */
-};
-
-struct backing_file_security_struct {
-	u32 uf_sid; /* top-level user file fsec->sid */
-	u32 layer_count; /* number of intermediate backing files */
-	struct backing_file_security_layer *layers;
-};
-
 struct superblock_security_struct {
 	u32 sid; /* SID of file system superblock */
 	u32 def_sid; /* default SID for labeling */
@@ -202,13 +190,6 @@ static inline struct file_security_struct *selinux_file(const struct file *file)
 	return file->f_security + selinux_blob_sizes.lbs_file;
 }
 
-static inline struct backing_file_security_struct *
-selinux_backing_file(const struct file *backing_file)
-{
-	void *blob = backing_file_security(backing_file);
-	return blob + selinux_blob_sizes.lbs_backing_file;
-}
-
 static inline struct inode_security_struct *
 selinux_inode(const struct inode *inode)
 {
-- 
2.50.1

